Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean

```diff
@@ -0,0 +1,311 @@
+
+* `TensorProduct.map_ker` computes the kernel of `TensorProduct.map g g'`
+in the presence of two short exact sequences.
+
+The proofs are those of Bourbaki, Algèbre, chap. 2, §3, n°6
+
+## TODO
+
+Add
+
+* Special case for short exact sequences of the form `M → N → N/M → 0`
+
+* Analogue for morphisms of algebras
+
+* Treat the noncommutative case
+
+* Treat the case of semirings (need to know what a short exact sequence is)
+
+-/
+
+
+namespace AddMonoidHom
+
+def Exact {M N P : Type _} [AddCommGroup M] [AddCommGroup N] [AddCommGroup P]
+  (f : outParam (M →+ N)) (g : outParam (N →+ P)) :=
+  ∀ y, y ∈ g.ker ↔ y ∈ f.range
+
+end AddMonoidHom
+
+namespace LinearMap
+
+open AddMonoidHom LinearMap
+
+variable (R : Type _) [CommRing R]
+variable {M N P : Type _} [AddCommGroup M] [AddCommGroup N][AddCommGroup P]
+  [Module R M] [Module R N] [Module R P]
+
+variable (f : M →ₗ[R] N) (g : N →ₗ[R] P)
+
+variable {R}
+def Exact := LinearMap.range f = LinearMap.ker g
+
+namespace Exact
+
+variable {f g}
+lemma comp_eq_zero (h : Exact f g) : g.comp f = 0 := by
+  ext x
+  simp only [coe_comp, Function.comp_apply, zero_apply]
+  rw [← mem_ker, ← h]
+  apply mem_range_self
+
+section rTensor
+
+
+open scoped TensorProduct
+
+example (N' : Submodule R N) : N' →ₗ[R] N := by
+  exact Submodule.subtype N'
+
+example (N' : Submodule R N) : Submodule R (N ⊗[R] P) :=
+  LinearMap.range (rTensorHom P (N'.subtype))
+
+
+variable (Q : Type _) [AddCommGroup Q] [Module R Q]
+
+variable (q : Q)
+
+example (R : Type _) [CommSemiring R]
+    (M N P : Type _) [AddCommMonoid M] [AddCommMonoid N] [AddCommMonoid P]
+      [Module R M] [Module R N] [Module R P]
+      (f : M →ₗ[R] N →ₗ[R] P) (m : M) :
+  N →ₗ[R] P := f m
+
+example (R : Type _) [CommSemiring R]
+    (M N P : Type _) [AddCommMonoid M] [AddCommMonoid N] [AddCommMonoid P]
+      [Module R M] [Module R N] [Module R P]
+      (f : M →ₗ[R] N →ₗ[R] P) (n : N) :
+  M →ₗ[R] P := f.flip n
+
+variable {Q} (g)
+lemma le_comap_range_rTensor (q : Q) :
+    LinearMap.range g ≤ (LinearMap.range (rTensor Q g)).comap ((TensorProduct.mk R P Q).flip q) := by
+  rintro x ⟨n, rfl⟩
+  exact ⟨n ⊗ₜ[R] q, rfl⟩
+
+variable (Q) {g}
+
+variable (hg : Function.Surjective g) (hfg : Exact f g)
+
+theorem rTensor.Surjective : Function.Surjective (rTensor Q g) := by
+  intro z
+  induction z using TensorProduct.induction_on with
+  | C0 => use 0; rw [map_zero]
+  | C1 p q =>
+      obtain ⟨n, rfl⟩ := hg p
+      use n ⊗ₜ[R] q
+      simp only [rTensor_tmul]
+  | Cp x y hx hy =>
+      obtain ⟨x, rfl⟩ := hx
+      obtain ⟨y, rfl⟩ := hy
+      use x + y
+      rw [map_add]
+
+private noncomputable
+def rTensor.inverse_ofRightInverse {h : P → N} (hgh : Function.RightInverse h g) :
+  P ⊗[R] Q →ₗ[R] N ⊗[R] Q ⧸ (LinearMap.range (rTensor Q f)) := by
+  apply TensorProduct.lift
+  apply LinearMap.mk₂ R (fun p q ↦ Submodule.Quotient.mk (h p ⊗ₜ[R] q))
+  . intro p p' q
```

Comment:
```suggestion
  · intro p p' q
```
and below, will fix the errors.

Reply:
Thanks ! (I definitely hate it that \cdot, \centerdot and . give seemingly looking results while only one of them is admitted.)